content/docs/guides/testing/_index.md

```diff
@@ -0,0 +1,82 @@
+
+## Unit Testing
+
+If you ever created unit tests for your application code, you can directly transfer your skills to testing Pulumi programs. Unit tests evaluate the behavior of your code in isolation, while all external dependencies are replaced by **test doubles** (mocks).
+
+Unit tests run in memory without any out-of-process calls, which makes them blazingly fast. Therefore, unit tests are suitable for fast feedback loops during development, including **Test-Driven Development** (TDD).
+
+You author the unit tests in the same language as the Pulumi program under test. You are free to use your favorite test and mock frameworks like Mocha for Node.js or NUnit for .NET.
+
+Because there are no real cloud resources created, you can't write a test that would evaluate the behavior of infrastructure. For example, you can't make HTTP requests to endpoints: there's no webserver to serve them.
+
+[Learn more and get started with Unit Testing]({{< relref "./unit" >}}).
+
+## Integration Testing
+
+Integration testing takes the opposite approach: the tests deploy real cloud resources and validate their actual **behavior**.
+
+The flow goes like this:
+
+1. A test invokes the Pulumi command-line interface (CLI) to deploy infrastructure to an ephemeral environment.
+2. When deployment succeeds, the test queries stack outputs to retrieve endpoints of the infrastructure: usually, a URL or a public IP address.
+3. The test verifies that the infrastructure behaves as expected: for example, it expects a valid HTML document from a health-check endpoint.
+4. Upon completion, the infrastructure is destroyed.
+
+The great advantage of integration tests is the ability to test the actual cloud infrastructure and its real properties. However, compared to unit tests, integrations tests take a long time to execute.
+
+Depending on the type of resources and frequency of testing, even short-lived ephemeral environments may incur notable charges from the cloud provider. Be sure to plan accordingly and measure frequently.
+
+[Learn more and get started with Integration Testing]({{< relref "./integration" >}}).
+
+## Testing with Policy as Code
+
+[Policy as Code]({{< relref "../crossguard" >}}) (also known as "CrossGuard") is Pulumi's offering to set guardrails and enforce compliance for cloud resources. In addition to authoring company-wide policies, CrossGuard enables another type of infrastructure testing.
+
+Policy-based tests run inside the Pulumi CLI before and after infrastructure provisioning. In contrast to "black-box" integration testing, policy rules have access to all input and output values of all cloud resources in the stack. As opposed to unit testing, policy rules can evaluate real values returned from the cloud provider instead of test doubles.
+
+Policy-based tests run against an ephemeral cloud environment, so the considerations of speed and cost still apply.
```

Comment:
This isn't quite true - these tests typically would run against a "real" long-lived cloud environment, like a staging or production environment.  The key differences from "integration tests" is that these tests *don't* stand up an ephemeral environment to test, and instead run "inside" an existing deployment.

Reply:
In fact, this point may be worth pulling up into the chart:

"Test environment": "In-memory", "long-lived stack (e.g. production)", "ephemeral stack"

?

Reply:
Ah, this means we have different views on the suggested workflow. Your statement is true for "proper" policy as code but I thought that here we repurpose the same functionality to run them on ephemeral environments similar to integration tests.

In your view, if I'm a developer and want to have a suit of policy tests -- to make sure the quality of my code overall -- how and when will I run them? Would I have my personal long-lived development environment? How do I run them in CI/CD, for pull requests etc?

@justinvp @pgavlin @joeduffy What are your views on the workflow for PaC testing?

Reply:
> In your view, if I'm a developer and want to have a suit of policy tests -- to make sure the quality of my code overall -- how and when will I run them? Would I have my personal long-lived development environment?

Yes, or a staging environment you open PRs against which fail on previews if you break a test (and fail deployments if you merge).

> How do I run them in CI/CD, for pull requests etc?

You either enable global policies for a set of environments, or provide a local pack of project-specific tests (policies) along with the project and set up `--policy-pack` parameters to the `pulumi up` you run inside your CI and PRs.

>  I thought that here we repurpose the same functionality to run them on ephemeral environments similar to integration tests.

I suppose you could run these policies inside deployment of ephemeral environments too, but we already have good support for validation of ephemeral environments with integration tests.  The thing that policies uniquely provide is tests that run intertwined with previews and updates against actual existing infrastructure.

Reply:
Isn't it important to be able to validate policies against ephemeral environments to make sure that, as you roll out to your *real* environments, you don't encounter unexpected violations?

/cc @clstokes

Reply:
@lukehoban What prevents me from using the same workflow for integration tests? I could have the same staging environment and run my integration tests against it?

> fail on previews

Previews aren't always enough: if I make a non-trivial change, I may not have enough outputs on a preview to validate. E.g. I can't check vnets of a cluster that isn't deployed because I just introduced it in my PR.

I would say that a unique thing about policies is access to all the resources and their interconnections, so it's a different level of validation.

Reply:
> Isn't it important to be able to validate policies against ephemeral environments to make sure that, as you roll out to your _real_ environments, you don't encounter unexpected violations?
> 
> /cc @clstokes

Yes, I think we would recommend validating policies against _all_ environments - ephemeral, long-lived testing environments, and production environments.

Reply:
@clstokes

> validating policies

and also to @ericrudder 's email remark

> compliance (or policy) verification?

This guide "repurposes" policies for developers (or DevOps) to run sort of more powerful unit tests. While they use exactly the same tech, their goals and workflows are different. They don't write company-wide restrictions, but they verify that a specific program works as expected in some policy-like aspects.

Does that make sense? Would this be clear to people? Does this change anything in the terminology and workflow?

Reply:
I mentioned both usage scenarios now